### DISCUSSION TITLE
Support building on MacOS without the need of MacOS SDK

### DISCUSSION THREAD

**Discussion by MadMcCrow** [godotengine/godot-proposals#10082] [Platforms]
MacOS SDK can only be obtained either through _Xcode_ or the _command line tools for Xcode_. These libraries cannot be redistributed due to their proprietary nature. It makes Godot entirely dependent on a non open, difficult to obtain (you have to sign up to Apple with a developer account, agreeing to their licence terms) subject to change piece of software.

I've been using nix on macOS to build, share, and distribute open-source code. This situation prevents having Godot on nix for Darwin. In the meantime stop gaps solutions exist (using pre-built binaries - incompatible with custom modules), accepting apple terms and conditions and building things manually, or stop using Godot and prefer a fully free development kit like raylib (it only depends on cocoa and carbon, both available through gnustep). 

I know removing the SDK entirely would probably mean having to drop the iOS target, and remove certain macOS integration, but having the Apple SDK as an optional dependency could be a solution (and you could have that option on by default on macOS, using Scons flag).

This change would be greatly appreciated by everyone who desires to fight for open platforms !

**Comment by AThousandShips** [comment]
Some concrete ideas on how to integrate alternative libraries would really help, otherwise this is very vague and hypothetical, especially what it would affect and what would be sacrificed 

Though I'd say if one wants a fully open platform one should use Linux instead, running on macOS is already limiting that, I'd say developing on/for a closed platform comes with those limitations, and using an open OS gets rid of all that

**Reply by MadMcCrow** [reply]
> Some concrete ideas on how to integrate alternative libraries would really help,

For that it would be necessary to know why the apple SDK is used in the first place. So far, I've noticed Godot uses these frameworks :
- FAPPKIT,
- FCOCOA,
- FCARBON,
- FAUDIOUNIT,
- FCOREAUDIO,
- FCOREMIDI,
- FIOKIT,
- FGAMECONTROLLER,
- FCOREHAPTICS,
- FCOREVIDEO,
- FAVFOUNDATION,
- FCOREMEDIA,
- FQUARTZCORE,
- FSECURITY,

How those are necessary, what could be made conditional? What could rely on older versions of the SDK (that can be downloaded without registering to apple, ie at most macOS SDK 11 ) ?

**Reply by AThousandShips** [reply]
I don't know almost anything about the internals of the macOS builds, but the question is really IMO if adding an option to run with alternative packages is:
* Reasonably easy and not adding a lot of extra work or maintenance cost/tech debt just to support it
* Valid and won't introduce risks for end users, I don't know how draconian Apple are about third party stuff on their OSs, but if using alternative packages runs the risk of having warranties etc. voided or keys invalidated on macOS or iOS then we need to be very careful about adding that as an option

**Reply by bruvzg** [reply]
> How those are necessary, what could be made conditional?

If you disable some engine features (audio i/o, webcam support), you can get rid of coreaudio and avfoundation, maybe few others, but the rest is required for the core engine functionality (window management, input, graphics). It's definitely impossible to narrow it down to cocoa/foundation only.

> What could rely on older versions of the SDK (that can be downloaded without registering to apple, ie at most macOS SDK 11 ) ?

Godot is targeting 10.15, while it is not tested with an older SDKs, it should be trivial to build it with 10.15 or newer SDK (by commenting some parts of code in `@available` but some features will be lost). Not sure where you have seen macOS 11 SDK downloadable without an account, AFAIK it was always locked behind account for any SDK version.

**Reply by MadMcCrow** [reply]
You can download the MacOS SDK version 11 here : 
`http://swcdn.apple.com/content/downloads/46/21/001-89745-A_56FM390IW5/v1um2qppgfdnam2e9cdqcqu2r6k8aa3lis/CLTools_macOSNMOS_SDK.pkg`
This means, you can (without breaking the terms and conditions) script its download and installation on any MacOS hardware.
this is how some packages are compiled on nixDarwin (nixpkgs for MacOS).

**Reply by bruvzg** [reply]
You can download 14.4 SDK (and likely any other version) the same way - https://swcdn.apple.com/content/downloads/14/48/052-59890-A_I0F5YGAY0Y/p9n40hio7892gou31o1v031ng6fnm9sb3c/CLTools_macOSNMOS_SDK.pkg

For the reference, current catalog file with all URLs seems to be - https://swscan.apple.com/content/catalogs/others/index-14-13-12-10.16-10.15-10.14-10.13-10.12-10.11-10.10-10.9-mountainlion-lion-snowleopard-leopard.merged-1.sucatalog.gz
